codalab/server/bundle_manager.py

```diff
@@ -299,28 +299,49 @@ def _bring_offline_stuck_running_bundles(self, workers):
                 logger.info('Bringing bundle offline %s: %s', bundle.uuid, failure_message)
                 self._model.transition_bundle_worker_offline(bundle)
 
-    def _schedule_run_bundles_on_workers(self, workers, user_owned):
+    def _schedule_run_bundles_on_workers(self, workers, staged_bundles_to_run):
         """
         Schedules STAGED bundles to run on the given workers. If user_owned is
         True, then schedules on workers run by the owner of each bundle.
         Otherwise, uses CodaLab-owned workers, which have user ID root_user_id.
         """
-        for bundle in self._model.batch_get_bundles(state=State.STAGED, bundle_type='run'):
-            if user_owned:
-                workers_list = workers.user_owned_workers(bundle.owner_id)
-            else:
+        # Reorder the stage_bundles so that bundles which were requested to run on a personal worker
+        # will be scheduled to run first
+        staged_bundles_to_run.sort(
+            key=lambda b: (b[0].metadata.request_queue is not None, b[0].metadata.request_queue),
+            reverse=True,
+        )
+
+        for bundle, bundle_resources in staged_bundles_to_run:
+            # Get user_owned workers.
+            private_workers = workers.user_owned_workers(bundle.owner_id)
+
+            # If there is no user_owned worker, try to schedule the current bundle to run on a CodaLab's public worker.
+            if not private_workers:
+                # Check if there is enough parallel run quota left for this user
                 if not self._model.get_user_parallel_run_quota_left(bundle.owner_id):
                     logger.info(
                         "User %s has no parallel run quota left, skipping job for now",
                         bundle.owner_id,
                     )
-                    continue  # Don't start this bundle yet
-                workers_list = workers.user_owned_workers(self._model.root_user_id)
-
-            workers_list = self._deduct_worker_resources(workers_list)
-            bundle_resources = self._compute_bundle_resources(bundle)
-            workers_list = self._filter_and_sort_workers(workers_list, bundle, bundle_resources)
+                    # Don't start this bundle yet, as there is no parallel_run_quota left for this user.
+                    continue
```

Comment:
Otherwise, I can go over my run quota like the following:
Say there's a parallel run quota of 1, and  I already have a running bundle. I can start a dummy private worker with 100mb of ram, and then submit a job requesting 4gb of ram. It would pass this check because I have a nonempty list of private workers, but then later on that private worker would get filtered away, and my run would get scheduled to the public workers. I could keep submitting as many runs as I want like that.